### DISCUSSION TITLE
How to use threads/atomics properly?

### DISCUSSION THREAD

**Discussion by aalhitennf** [emilk/egui#1428] [Q&A] [answered]
I'm trying to implement app that has background thread, which updates some atomic variables, and shows them in labels. I can make it work, but in the gui, values update only when i move mouse or press keyboard buttons. Is there a way to force egui to update i.e. every 500ms?

My project is modified eframe_template for quick testing. Basically what i have is pub static AtomicI32, i start thread that increments the atomic every second, then i create the egui app, and use and show that atomic in label, but it updates only if i move the mouse cursor or press keyboard keys.

**Comment by aalhitennf** [comment]
I also tried wrapping I32 inside Arc and Mutex but same behaviour continues.

**Comment by coderedart** [comment, marked as answer]
So, you get this frame in your update closure which you can clone and use to trigger request repaint,  that you can call from another thread if you update the value or after 500 ms or whatever condition you prefer.  

https://docs.rs/epi/0.17.0/epi/struct.Frame.html#method.request_repaint

**Reply by coderedart** [reply]
to expand open this, egui only draws a frame IF and only IF there's an event. this is done as a power saving measure i assume. the other alternative is to draw every frame (to match vsync), 60 fps in general. which is overkill for a lot of desktop applications unless you are a game engine or something. 

so, it provides you with a Frame struct on which you can call request_repaint from another thread if you want to update egui display because there's an update on the underlying data. it can work in multiple ways, but consider request_repaint as sending a "fake" window event which makes egui receive the event on main thread and run the update closure.

**Reply by aalhitennf** [reply]
I spawned my bg thread in my apps setup function, cloned the frame for it and call request_repaint with duration. Now it updated the gui as i wanted. Thanks a lot!
